From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Tue, 20 Feb 2024 18:24:16 -0800
Subject: [PATCH] Fix entity tracker desync when new players are added to the
 tracker

The delta position packet instructs the client to update
the entity position by a position difference. However, this position
difference is relative to the last position in the entity tracker
state, not the last position which has been sent to the player. As
a result, if the last position the player has recorded is different
than the one stored in the entity tracker (which occurs when a new
player is added to an existing entity tracker state) then the sent
position difference will cause a position desync for the client.

We can resolve this problem by either tracking the last position
sent per-player, or by simply resetting the last sent position
in the entity tracker state every time a new player is added.
Resetting the last sent position every time a new player is
added to the tracker is just easier to do, so that is what
this patch does.

This patch also fixes entities appearing to disappear when
teleporting to players by changing the initial position
in the spawn packet to the entities current tracking position.
When teleporting, the spawn packet will contain the old position
which is most likely in an unloaded chunk - which means that the
client will not tick the entity and thus not lerp the entity
from its old position to its new position.

diff --git a/net/minecraft/network/protocol/game/ClientboundAddEntityPacket.java b/net/minecraft/network/protocol/game/ClientboundAddEntityPacket.java
index cff05f64af7f425b77cbbd7680c5cc592faa798c..0ab4080c8e30d5ee14ce5fcbac64b0316ea6a949 100644
--- a/net/minecraft/network/protocol/game/ClientboundAddEntityPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundAddEntityPacket.java
@@ -38,9 +38,11 @@ public class ClientboundAddEntityPacket implements Packet<ClientGamePacketListen
         this(
             entity.getId(),
             entity.getUUID(),
-            serverEntity.getPositionBase().x(),
-            serverEntity.getPositionBase().y(),
-            serverEntity.getPositionBase().z(),
+            // Paper start - fix entity tracker desync
+            entity.trackingPosition().x(),
+            entity.trackingPosition().y(),
+            entity.trackingPosition().z(),
+            // Paper end - fix entity tracker desync
             serverEntity.getLastSentXRot(),
             serverEntity.getLastSentYRot(),
             entity.getType(),
diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index 206a897e9d74b48d38e2ea1652e9b7ed4735f828..5a8f8eb8f6400ab2397483400ef3853fd0cf6895 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -1378,6 +1378,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
                         ChunkMap.this.level.debugSynchronizers().startTrackingEntity(player, this.entity);
                         }
                         // Paper end - entity tracking events
+                        this.serverEntity.onPlayerAdd(); // Paper - fix desync when a player is added to the tracker
                     }
                 } else {
                     this.removePlayer(player);
diff --git a/net/minecraft/server/level/ServerEntity.java b/net/minecraft/server/level/ServerEntity.java
index 267259a79c7668da85d206f9dcbde16b379e2738..84a0c43bdaf7c6fe0067a421ba39964b36e679a5 100644
--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -102,6 +102,13 @@ public class ServerEntity {
         this.trackedDataValues = entity.getEntityData().getNonDefaultValues();
     }
 
+    // Paper start - fix desync when a player is added to the tracker
+    private boolean forceStateResync;
+    public void onPlayerAdd() {
+        this.forceStateResync = true;
+    }
+    // Paper end - fix desync when a player is added to the tracker
+
     public void sendChanges() {
         // Paper start - optimise collisions
         if (((ca.spottedleaf.moonrise.patches.chunk_system.entity.ChunkSystemEntity)this.entity).moonrise$isHardColliding()) {
@@ -147,7 +154,7 @@ public class ServerEntity {
             this.entity.syncPosition = false;
         }
 
-        if (this.entity.needsSync || this.updateInterval.test(this.tickCount) || this.entity.getEntityData().isDirty()) {
+        if (this.forceStateResync || this.entity.needsSync || this.updateInterval.test(this.tickCount) || this.entity.getEntityData().isDirty()) { // Paper - fix desync when a player is added to the tracker
             byte yRotn = Mth.packDegrees(this.entity.getYRot());
             byte xRotn = Mth.packDegrees(this.entity.getXRot());
             boolean shouldSendRotation = Math.abs(yRotn - this.lastSentYRot) >= 1 || Math.abs(xRotn - this.lastSentXRot) >= 1;
@@ -167,18 +174,38 @@ public class ServerEntity {
             } else {
                 this.teleportDelay++;
                 PositionPath position = this.interpolationTracker.getPositionPath(currentPosition);
-                this.interpolationTracker.clear();
+
+                // Paper start - reduce allocation of Vec3 here
+                Vec3 base = this.positionCodec.base;
+                double dx = currentPosition.x - base.x;
+                double dy = currentPosition.y - base.y;
+                double dz = currentPosition.z - base.z;
+                boolean positionChanged = (dx * dx + dy * dy + dz * dz) >= 7.62939453125E-6D;
+                // Paper end - reduce allocation of Vec3 here
+
                 MovementPacket<ClientGamePacketListener> packet;
-                if (this.entity instanceof AbstractArrow) {
-                    packet = this.createMovePacket(position, yRotn, xRotn, true, true);
+                boolean pos = positionChanged || this.tickCount % 60 == 0;
+                long xa = this.positionCodec.encodeX(currentPosition);
+                long ya = this.positionCodec.encodeY(currentPosition);
+                long za = this.positionCodec.encodeZ(currentPosition);
+                boolean deltaTooBig = xa < -32768L || xa > 32767L || ya < -32768L || ya > 32767L || za < -32768L || za > 32767L;
+                if (this.forceStateResync || this.entity.getRequiresPrecisePosition() // Paper - fix desync when a player is added to the tracker
+                    || deltaTooBig
+                    || this.teleportDelay > 400
+                    || this.wasRiding
+                    || this.wasOnGround != this.entity.onGround()) {
+                    this.wasOnGround = this.entity.onGround();
+                    this.teleportDelay = 0;
+                    packet = ClientboundEntityPositionSyncPacket.of(this.entity);
+                } else if ((!pos || !shouldSendRotation) && !(this.entity instanceof AbstractArrow)) {
+                    if (pos) {
+                        packet = this.createMovePacket(position, yRotn, xRotn, true, true);
+                    } else if (shouldSendRotation) {
+                        packet = this.createMovePacket(position, yRotn, xRotn, false, true);
+                    } else {
+                        packet = null;
+                    }
                 } else {
-                    // Paper start - reduce allocation of Vec3 here
-                    Vec3 base = this.positionCodec.base;
-                    double dx = currentPosition.x - base.x;
-                    double dy = currentPosition.y - base.y;
-                    double dz = currentPosition.z - base.z;
-                    boolean positionChanged = (dx * dx + dy * dy + dz * dz) >= 7.62939453125E-6D;
-                    // Paper end - reduce allocation of Vec3 here
                     boolean shouldSendPosition = positionChanged || this.tickCount % 60 == 0;
                     packet = this.createMovePacket(position, yRotn, xRotn, shouldSendPosition, shouldSendRotation);
                 }
@@ -229,6 +256,7 @@ public class ServerEntity {
             }
 
             this.entity.needsSync = false;
+            this.forceStateResync = false; // Paper - fix desync when a player is added to the tracker
         }
 
         this.tickCount++;
